### PR TITLE
Add Detekt to the build

### PR DESCRIPTION
**Pull request by RenFraser**
Adds [Detekt](https://github.com/detekt/detekt), a popular open-source static analysis tool to the build.

I've set the build task to depend on running Detekt, rather than having it as a GitHub action. This should catch issues earlier in the development cycle. I've also set the maximum number of issues to be equal to the number of issues that we have at the moment when using the default ruleset. I think it'd be good to start off with the default rules and override them only if they become noise.

### CODE REVIEW

**Comment by themkat** [comment]
I would very much like to introduce Detekt too! 🙂 It is an awesome analysis tool. A question from me: Would it make sense to maybe run Detekt analysis on the `shared`-module as well?  


Before I think about merging, I would like to hear @fwcd s opinions here. Based on earlier discussions, this is something it seems FW could have some opinions on 🙂

**Comment by RenFraser** [comment]
> on the shared-module as well?

I wasn't too sure if the shared module was held to the same standard but definitely happy to add it to this one too! In the future, it might be a good idea to extract out logic like this in the build script and apply it as a plugin. If we were to do that we could have our logic as a plugin and do something like:

```kotlin
plugins {
    id("ourReusableDetektBuildLogic")
}
```

**Comment by fwcd** [comment]
It's a great idea and I think having it run in CI too would not hurt either.

That said, I'm not sure how to do this well without "mass reformatting PRs", which are hard to review, add noise to the commit history and are a bit messy in general, even if we choose to ignore most rules for now. Other projects (such as [`mixxx`](https://github.com/mixxxdj/mixxx)) have faced similar issues when introducing linters and formatters into their already very large codebase. `clang-format` solves this by only requiring the diffs of newly changed code to follow the formatting guidelines, thereby allowing more gradual adoption throughout the codebase. Perhaps Detekt supports something similar?

**Comment by RenFraser** [comment]
That's fair. I'm not sure how we usually do it here but I've always liked the idea of adding a formatter or linter in one commit then running some sort of automated command and committing the diff, without manually changing anything. Though that also relies on a lot of trust in the open-source community. Perhaps if we're going to address the issues the tool finds, we could do it in much smaller PRs.

I don't want to propose that we do a pass over the code with it, which is why I've set the following in the config:

```yaml
maxIssues: 977 # set so that we don't immediately break the build
```

By doing this, we're saying "If any more errors are introduced, break the build". So it's not running on git's changed files explicitly, but it is doing it a slightly different way. In their documentation, Detekt provides [a page on running it as a git hook.](https://detekt.dev/docs/gettingstarted/git-pre-commit-hook/) We could also consider this if you see the value in it. 

> I think having it run in CI too would not hurt either.

Sounds good. I've added a step in the build job.

I've pushed a new revision and squashed it into the original commit. I've 

- Added a step to the build job to run detekt. 
- Added it to the root `check` task dependencies. 
- Refactored the changes out and placed them in the root dir so that it runs on both shared and server source sets. 
  - It will also run against the grammar subproject but there is only the `build.gradle.kts` file in there. 
- Raised the maximum issues to 977 to encompass the shared subproject.
- Added a configuration to run detekt on subprojects in parallel to speed up the processing time.
- Rebased on to the latest main branch.

I'd also like to refactor the detekt logic out into a project plugin at some point but will wait until #424 is merged first.

**Comment by RenFraser** [comment]
With this change, we could close:

- #342 
- #304

**Comment by themkat** [comment]
> It's a great idea and I think having it run in CI too would not hurt either.
> 
> That said, I'm not sure how to do this well without "mass reformatting PRs", which are hard to review, add noise to the commit history and are a bit messy in general, even if we choose to ignore most rules for now. Other projects (such as [`mixxx`](https://github.com/mixxxdj/mixxx)) have faced similar issues when introducing linters and formatters into their already very large codebase. `clang-format` solves this by only requiring the diffs of newly changed code to follow the formatting guidelines, thereby allowing more gradual adoption throughout the codebase. Perhaps Detekt supports something similar?


If I'm not mistaken, isn't [Detekt Baseline](https://detekt.dev/docs/introduction/baseline) exactly that? In my projects at work and home, I have never needed it, so had read up and experiment a bit. To my understanding, we might be able to skip the `maxIssues` property by creating a baseline. Then it should only report on new issues and code that is touched. 


Example on config for having a baseline for the regular detekt run, plus a special task to create a baseline when needed:
```kotlin
detekt {
    allRules = false // activate all available (even unstable) rules.
    buildUponDefaultConfig = true // preconfigure defaults
    parallel = true
    config = files("$rootDir/detekt.yml")
    baseline = file("$rootDir/detekt_baseline.xml")
    source = files(projectDir)
}

tasks.register<io.gitlab.arturbosch.detekt.DetektCreateBaselineTask>("createDetektBaseline") {
    description = "Overrides current baseline."
    buildUponDefaultConfig.set(true)
    ignoreFailures.set(true)
    parallel.set(true)
    setSource(files(projectDir))
    config.setFrom(files("$rootDir/detekt.yml"))
    baseline.set(file("$rootDir/detekt_baseline.xml"))
    include("**/*.kt")
    include("**/*.kts")
    exclude("shared/build/**/*.*")
    exclude("server/build/**/*.*")
}
```
I noticed that the Gradle build-directories were included (even on the standard run before my changes), so had to manually exclude the files in them. (to avoid issues with the kts-scripts in the shared module when they are put in build during building)

**Comment by themkat** [comment]
> That sounds great @themkat. Sounds like it's the proper way to do what I've done from new changes instead of setting a max issues.
> 
> To confirm, is the idea that we generate a baseline, and reference it in the detekt block and then any following detekt runs will only fail if there's new issues that are not found in the baseline xml?

As I understand it, yes 🙂 If I create a baseline with the code above, and then run regular `gradle detekt`, then no errors occur. We can also generate a new baseline if we ever needed it, which could be when we fix some older issues.

**Comment by RenFraser** [comment]
As an aside, I'm pushing squashed commits. Is it possible to push commits addressing feedback and then squash them into one before merging? I haven't used GitHub as much as I'd like and know it's painful to review squashed commit diffs between revisions

**Comment by RenFraser** [comment]
> > That sounds great @themkat. Sounds like it's the proper way to do what I've done from new changes instead of setting a max issues.
> 
> > 
> 
> > To confirm, is the idea that we generate a baseline, and reference it in the detekt block and then any following detekt runs will only fail if there's new issues that are not found in the baseline xml?
> 
> 
> 
> As I understand it, yes 🙂 If I create a baseline with the code above, and then run regular `gradle detekt`, then no errors occur. We can also generate a new baseline if we ever needed it, which could be when we fix some older issues.

Ok that sounds perfect. I'll make the necessary adjustment in the next revision over the next day or two. Thanks for the info!

**Comment by RenFraser** [comment]
@themkat I've added the baseline detekt task and generated the baseline file, re-run detekt with the max-issues removed and it's worked perfectly. I'm glad that you suggested it because I've learned something new here. :)

I've used a separate commit instead of squashing it because it looks like you can squash-merge in GitHub (still a noob sorry) and hopefully the separate commits will make review easier.

**Review by themkat — APPROVED** [review]
Awesome work @RenFraser ! Will give @fwcd a few days to have a second look, just in case there is something I'm missing 🙂

**Comment by themkat** [comment]
Been a few days now, so.... Let's merge this and see how well it works on new PRs 🙂 

Thank you @RenFraser for all your contributions lately! (issues, discussions, PRs). It has made this project more fun to work with 🙂

**Review by fwcd — COMMENTED** [review]
Thanks to both of you for looking into this!

File: .github/workflows/build.yml
Comment (RenFraser) [reply .github/workflows/build.yml:26]: You're right @fwcd, it's duplicated because the `build` task depends on `check`, which depends on `detekt`. I'm wondering how we should go about this now. It's great to be more explicit about the steps that we're performing in the build as a part of our CI checks, but I also see a lot of value in keeping the `check` task dependent on the static analysis detekt performs. Do we know what any other projects do (or if there's a best practice around this)? If not I'll take a look at some popular libraries on GitHub, like ktlint, and see what they do.

Comment (fwcd) [reply .github/workflows/build.yml:26]: What I meant is that

```
- uses: gradle/gradle-build-action@v2
```

is a fully distinct step from

```
- name: Build
   run: ./gradlew :server:build :shared:build -PjavaVersion=${{ matrix.java }}
```

and we probably don't need to run the former one twice. In fact, we might not even need it at all, since the Gradle wrapper should already take care of installing Gradle.

(yes, YAML syntax might be a bit subtle here)

Comment (RenFraser) [reply .github/workflows/build.yml:26]: Ah I see what you mean now. 

Should we do:

```
- name: Build
   run: ./gradlew build -PjavaVersion=${{ matrix.java }}
```

To capture the root level configurations as well? 
Is there a reason we don't build the `grammar` subproject on the pipeline? I haven't seen the discussion about it just yet

Comment (fwcd) [reply .github/workflows/build.yml:26]: The `grammar` subproject doesn't really need to be built, since it's just a bunch of JSON files. The reason we have a Gradle subproject for it is to have a convenient `distZip` task for packaging up releases.

**Pull request merged**
